### PR TITLE
Finished database implementation.

### PR DESCRIPTION
**Pull request by teamagam3**
* Added entities to cache the data.
* Added `EntityMapper`, to map between `Room` objects to ours.

### CODE REVIEW

File: app/src/main/java/com/teamagam/gimelgimel/app/injectors/modules/RepositoryModule.java [hunk 1]
```diff
@@ -212,4 +216,10 @@ NewMessageIndicationRepository provideNewMessageIndicationRepository(
       NewMessageIndicationDataRepository newMessageIndicationDataRepository) {
     return newMessageIndicationDataRepository;
   }
+
+  @Provides
+  @Singleton
+  MessagesCache provideMessagesCache(MessagesDao dao, MessagesEntityMapper converter) {
+    return new MessagesDataCache(dao, converter);
```
Comment (teamagam5) [inline app/src/main/java/com/teamagam/gimelgimel/app/injectors/modules/RepositoryModule.java]: converter -> mapper ?

File: data/src/main/java/com/teamagam/gimelgimel/data/config/Constants.java [hunk 2]
```diff
@@ -11,7 +11,6 @@
   public static final String MESSAGING_SERVER_DEFAULT = MESSAGING_SERVER_URL_1;
   public static final long CONNECTION_SERVER_TIME_OUT_SECONDS = 120;
 
-  //Proxy
```
Comment (teamagam5) [inline data/src/main/java/com/teamagam/gimelgimel/data/config/Constants.java:14]: U know how much I like removing comments. But I admit that in this case I'm not sure whether or not it's needed.

File: data/src/main/java/com/teamagam/gimelgimel/data/location/repository/UsersLocationDataRepository.java [hunk 3]
```diff
@@ -1,45 +1,42 @@
...
-import java.util.Stack;
-import java.util.TreeMap;
 import javax.inject.Inject;
 import javax.inject.Singleton;
 
 @Singleton
 public class UsersLocationDataRepository implements UsersLocationRepository {
 
-  private Map<String, Stack<LocationSample>> mUsersLocations;
+  private UserLocationDao mUserLocationDao;
```
Comment (teamagam5) [inline data/src/main/java/com/teamagam/gimelgimel/data/location/repository/UsersLocationDataRepository.java]: Be consistent with your naming. In VectorLayersDataRepository u named it just "mDao".

File: data/src/main/java/com/teamagam/gimelgimel/data/location/repository/UsersLocationDataRepository.java [hunk 4]
```diff
@@ -1,45 +1,42 @@
...
-    mUsersLocations.get(userId).add(userLocation.getLocationSample());
+    mUserLocationDao.insertUserLocation(mMapper.convertToEntity(userLocation));
   }
 
   @Override
   public Iterable<UserLocation> getLastLocations() {
-    List<UserLocation> list = new ArrayList<>();
-    for (Map.Entry<String, Stack<LocationSample>> entry : mUsersLocations.entrySet()) {
-      list.add(extractLastLocation(entry));
+    List<UserLocationEntity> lastLocations = mUserLocationDao.getLastLocations();
```
Comment (teamagam5) [inline data/src/main/java/com/teamagam/gimelgimel/data/location/repository/UsersLocationDataRepository.java]: lastLocations -> lastLocationsEntities ?

File: data/src/main/java/com/teamagam/gimelgimel/data/map/adapter/GeometryDataMapper.java [hunk 5]
```diff
@@ -63,6 +62,18 @@ public PointGeometry transform(Point geoData) {
     return createPointGeometry(lat, lon, alt);
   }
 
+  public Geometry transformBySuperclass(com.teamagam.geogson.core.model.Geometry geoData) {
```
Comment (teamagam5) [inline data/src/main/java/com/teamagam/gimelgimel/data/map/adapter/GeometryDataMapper.java:65]: transformToDomainBySuperclass ?

File: data/src/main/java/com/teamagam/gimelgimel/data/message/repository/cache/room/dao/MessagesDao.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+  List<ChatMessageEntity> getMessages();
+
+  @Query("SELECT * FROM messages ORDER BY creation_date DESC LIMIT 1")
+  Flowable<ChatMessageEntity> getLatestMessage();
+
+  @Query("SELECT * FROM messages WHERE messageId = :id")
+  ChatMessageEntity getMessageById(String id);
+
+  @Query("SELECT * FROM messages ORDER BY creation_date DESC LIMIT 1")
+  ChatMessageEntity getLastMessage();
```
Comment (teamagam5) [inline data/src/main/java/com/teamagam/gimelgimel/data/message/repository/cache/room/dao/MessagesDao.java]: Move this to be near `getLatest...`
BTW is it used?

File: data/src/main/java/com/teamagam/gimelgimel/data/message/repository/cache/room/dao/UserLocationDao.java [hunk 7]
```diff
@@ -1,7 +1,24 @@
...
 
 import android.arch.persistence.room.Dao;
+import android.arch.persistence.room.Insert;
+import android.arch.persistence.room.Query;
+import android.arch.persistence.room.RoomWarnings;
+import com.teamagam.gimelgimel.data.message.repository.cache.room.entities.UserLocationEntity;
+import java.util.List;
 
 @Dao
 public interface UserLocationDao {
```
Comment (teamagam5) [inline data/src/main/java/com/teamagam/gimelgimel/data/message/repository/cache/room/dao/UserLocationDao.java:11]: UserLocation*s*Dao ?

File: data/src/main/java/com/teamagam/gimelgimel/data/message/repository/cache/room/mappers/EntityMapper.java [hunk 8]
```diff
@@ -0,0 +1,7 @@
+package com.teamagam.gimelgimel.data.message.repository.cache.room.mappers;
+
+public interface EntityMapper<T, R> {
+  T convertToDomain(R entity);
```
Comment (teamagam5) [inline data/src/main/java/com/teamagam/gimelgimel/data/message/repository/cache/room/mappers/EntityMapper.java]: `map...`?

File: data/src/main/java/com/teamagam/gimelgimel/data/message/repository/cache/room/mappers/UserLocationsEntityMapper.java [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+  public UserLocation convertToDomain(UserLocationEntity entity) {
+    LocationSampleEntity sampleEntity = entity.location;
+    return new UserLocation(entity.user, createLocationSample(sampleEntity));
+  }
+
+  public UserLocationEntity convertToEntity(UserLocation userLocation) {
+    LocationSample sample = userLocation.getLocationSample();
+    UserLocationEntity entity = new UserLocationEntity();
+
+    entity.id = UUID.randomUUID().toString();
```
Comment (teamagam5) [inline data/src/main/java/com/teamagam/gimelgimel/data/message/repository/cache/room/mappers/UserLocationsEntityMapper.java]: Why do u need to generate? Can't u use the user's ID for example?

File: dependencies.gradle [hunk 10]
```diff
@@ -23,6 +23,7 @@ ext {
 
   retrofitVersion = '2.2.0'
   powermockVersion = '1.7.0'
+  archVersion = '1.0.0-alpha3'
```
Comment (teamagam5) [inline dependencies.gradle:26]: What does `archVersion` mean?

**Review by teamagam5 — APPROVED** [review]
I think it's only renaming but go through it...

File: data/src/main/java/com/teamagam/gimelgimel/data/config/Constants.java (hunk 2 above)
Comment (teamagam3) [reply data/src/main/java/com/teamagam/gimelgimel/data/config/Constants.java:14]: The variable names are well explained.
We use blank line to separate the categories.

File: data/src/main/java/com/teamagam/gimelgimel/data/message/repository/cache/room/mappers/UserLocationsEntityMapper.java (hunk 9 above)
Comment (teamagam3) [reply data/src/main/java/com/teamagam/gimelgimel/data/message/repository/cache/room/mappers/UserLocationsEntityMapper.java]: No, it should be unique.

File: dependencies.gradle (hunk 10 above)
Comment (teamagam3) [reply dependencies.gradle:26]: architecture library version

**Pull request merged**